server/src/main/java/org/opensearch/common/settings/WriteableSetting.java

```diff
@@ -0,0 +1,273 @@
+        this.type = type;
+    }
+
+    /**
+     * Wrap a {@link Setting} read from a stream.
+     *
+     * @param in Input to read the value from.
+     * @throws IOException if there is an error reading the values
+     */
+    public WriteableSetting(StreamInput in) throws IOException {
+        // Read the type
+        this.type = in.readEnum(SettingType.class);
+        // Read the key
+        String key = in.readString();
+        // Read the default value
+        Object defaultValue = readDefaultValue(in);
+        // Read a boolean specifying whether the fallback settings is null
+        WriteableSetting fallback = null;
+        boolean hasFallback = in.readBoolean();
+        if (hasFallback) {
+            fallback = new WriteableSetting(in);
+        }
+        // We are using known types so don't need the parser
+        // We are not using validator
+        // Read properties
+        EnumSet<Property> propSet = in.readEnumSet(Property.class);
+        // Put it all in a setting object
+        this.setting = createSetting(type, key, defaultValue, fallback, propSet.toArray(Property[]::new));
+    }
+
+    /**
+     * Due to type erasure, it is impossible to determine the generic type of a Setting at runtime.
+     * All settings have a non-null default, however, so the type of the default can be used to determine the setting type.
+     *
+     * @param setting The setting with a generic type.
+     * @return The corresponding {@link SettingType} for the default value.
+     */
+    private static SettingType getGenericTypeFromDefault(Setting<?> setting) {
+        String typeStr = null;
+        try {
+            // This throws NPE on null default
+            typeStr = setting.getDefault(Settings.EMPTY).getClass().getSimpleName();
+            // This throws IAE if not in enum
+            return SettingType.valueOf(typeStr);
+        } catch (NullPointerException e) {
+            throw new IllegalArgumentException("Unable to determine the generic type of this setting with a null default value.");
+        } catch (IllegalArgumentException e) {
+            throw new UnsupportedOperationException("This class is not yet set up to handle the generic type: " + typeStr);
+        }
+    }
+
+    /**
+     * Gets the wrapped setting. Use {@link #getType()} to determine its generic type.
+     *
+     * @return The wrapped setting.
+     */
+    public Setting<?> getSetting() {
+        return this.setting;
+    }
+
+    /**
+     * Gets the generic type of the wrapped setting.
+     *
+     * @return The wrapped setting's generic type.
+     */
+    public SettingType getType() {
+        return this.type;
+    }
+
+    @SuppressWarnings("unchecked")
+    private Setting<?> createSetting(
+        SettingType genericType,
+        String key,
+        Object defaultValue,
+        WriteableSetting fallback,
+        Property[] propArray
+    ) {
+        switch (genericType) {
+            case Boolean:
```

Comment:
Instead of having the switch case here and to make the code a little cleaner, can we create a map something like:
`{int: Setting.intSetting(params)}` and then fetch the value from the map?

Reply:
Great idea, but I'm not sure it will work.
 - I'll need two maps, one for the fallbacks and one without (see the test class where I did exactly this)
   - And then I'd need to add the logic to test fallback first to determine which map to use
     - Except for the case with no fallback
 - In the future I expect a NamedWriteable type which will be a superclass
   - So I'd have to put in logic to check `instanceof` before looking it up... seems handling in `default` is easier here
 - I don't expect this list to grow very much and it seems to strike a good balance to me.